app/abci/handlers.go

```diff
@@ -192,8 +191,26 @@ func (h *Handlers) PrepareProposalHandler() sdk.PrepareProposalHandler {
 		}
 
 		var injection []byte
-		if req.Height > ctx.ConsensusParams().Abci.VoteExtensionsEnableHeight && collectSigs {
-			err := baseapp.ValidateVoteExtensions(ctx, h.stakingKeeper, req.Height, ctx.ChainID(), req.LocalLastCommit)
+		if VoteExtensionsEnabled(ctx) && collectSigs {
+			for i, vote := range req.LocalLastCommit.Votes {
+				// Verify the signatures since they're not guaranteed to have passed VerifyVoteExtension.
+				if err := h.verifyBatchSignatures(ctx, batch.BatchNumber, batch.BatchId, vote.VoteExtension, vote.Validator.Address); err != nil {
```

Comment:
Shouldn't we also check if the voter was in the previous batch's validator tree?

Reply:
Ah, I think I was mistaken. If a validator joins at height `H`, it would not be in the local last commit at height `H+1`.